### PR TITLE
Download Docker-Compose-v2 automatically if necessary

### LINKED ISSUE #62: [kwokctl] Download Docker-Compose-v2 automatically if necessary
https://github.com/docker/compose

### PR DESCRIPTION
**Pull request by fengshunli**
Fixes #62
Continue #87

### CODE REVIEW

File: pkg/kwokctl/vars/vars.go [hunk 1]
```diff
@@ -127,6 +129,10 @@ var (
 	// MustKubectlBinary is the binary of kubectl.
 	MustKubectlBinary = "https://dl.k8s.io/release/" + KubeVersion + "/bin/" + runtime.GOOS + "/" + runtime.GOARCH + "/kubectl" + BinSuffix
 
+	Arch = map[string]string{"arm64": "aarch64"}
+
```
Comment (wzshiming) [inline pkg/kwokctl/vars/vars.go:133]:
```suggestion
```

File: pkg/kwokctl/vars/vars.go [hunk 2]
```diff
@@ -278,3 +284,10 @@ func addPrefixV(version string) string {
 	}
 	return "v" + version
 }
+
+func archAlias(arch string) string {
+	if _, ok := Arch[arch]; ok {
+		return Arch[arch]
+	}
+	return arch
+}
```
Comment (wzshiming) [inline pkg/kwokctl/vars/vars.go:293]:
```suggestion
var archAliasMapping = map[string]string{
	"arm64": "aarch64",
}

func archAlias(arch string) string {
	if v, ok := archAliasMapping[arch]; ok {
		return v
	}
	return arch
}
```

Comment (wzshiming) [reply pkg/kwokctl/vars/vars.go:293]: Please add alias for amd, 386 arm, etc

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 3]
```diff
@@ -384,3 +399,25 @@ func (c *Cluster) ListImages(ctx context.Context, actual bool) ([]string, error)
...
 	}, nil
 }
+
+func (c *Cluster) getCompose(ctx context.Context, args ...string) ([]string, error) {
+	conf, err := c.Config()
+	if err != nil {
+		return nil, err
+	}
+	if conf.Runtime == "docker" {
+		cmd := exec.Command(conf.Runtime, "compose")
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:409]: Uniform change to exec.Command wrap `utils.Exec`

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 4]
```diff
@@ -384,3 +399,25 @@ func (c *Cluster) ListImages(ctx context.Context, actual bool) ([]string, error)
...
+		_, err = cmd.CombinedOutput()
+		if err != nil {
+			dockerComposePath := utils.PathJoin(conf.Workdir, "bin", "docker-compose")
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:412]:
```suggestion
			dockerComposePath := utils.PathJoin(conf.Workdir, "bin", "docker-compose"+vars.BinSuffix)
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Comment by wzshiming** [comment]
Don't open a PR every time and `git push -f` directly onto the original branch

**Comment by fengshunli** [comment]
> git push -f
ok

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 5]
```diff
@@ -21,7 +21,6 @@ import (
 	"fmt"
 	"io"
 	"os"
-
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:24]: Keep a blank line

File: pkg/kwokctl/vars/vars.go [hunk 6]
```diff
@@ -278,3 +282,15 @@ func addPrefixV(version string) string {
 	}
 	return "v" + version
 }
+
+var archMapping = map[string]string{
+	"arm64":  "aarch64",
+	"x86_64": "x86_64",
```
Comment (wzshiming) [inline pkg/kwokctl/vars/vars.go:288]:
```suggestion
	"arm64": "aarch64",
	"arm: "armv7",
	"amd64": "x86_64",
	"386": "x86",
```

File: pkg/kwokctl/vars/vars.go [hunk 7]
```diff
@@ -278,3 +282,15 @@ func addPrefixV(version string) string {
...
+}
+
+func archAlias(arch string) string {
+	if _, ok := archMapping[arch]; ok {
+		return archMapping[arch]
```
Comment (wzshiming) [inline pkg/kwokctl/vars/vars.go:293]:
```suggestion
	if v, ok := archMapping[arch]; ok {
		return v
```

File: pkg/kwokctl/vars/vars.go [hunk 8]
```diff
@@ -64,6 +64,8 @@ var (
 	// KubeVersion is the version of Kubernetes to use.
 	KubeVersion = addPrefixV(getEnv("KWOK_KUBE_VERSION", consts.KubeVersion))
 
+	DockerComposeVersion = addPrefixV(getEnv("KWOK_DOCKER_COMPOSE_VERSION", "v2.13.0"))
```
Comment (wzshiming) [inline pkg/kwokctl/vars/vars.go:67]:
```suggestion
	// DockerComposeVersion is the version of docker-compose to use.
	DockerComposeVersion = addPrefixV(getEnv("KWOK_DOCKER_COMPOSE_VERSION", "v2.13.0"))
```

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 9]
```diff
@@ -384,3 +398,23 @@ func (c *Cluster) ListImages(ctx context.Context, actual bool) ([]string, error)
...
+			dockerComposePath := utils.PathJoin(conf.Workdir, "bin", "docker-compose"+vars.BinSuffix)
+			err = utils.DownloadWithCache(ctx, conf.CacheDir, vars.MustDockerComposeBinary, dockerComposePath, 0755, conf.QuietPull)
+			if err != nil {
+				return nil, err
+			}
+			return append([]string{dockerComposePath}, args...), nil
+		}
+	}
+	return append([]string{conf.Runtime, "compose"}, args...), nil
+
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:419]:
```suggestion
```

**Review by wzshiming — CHANGES_REQUESTED** [review]
Delete the download docker-compose logic from CI for test the patch is working

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 10]
```diff
@@ -261,13 +260,21 @@ func (c *Cluster) Up(ctx context.Context) error {
...
 		return err
 	}
-	args := []string{"compose", "up", "-d"}
+
+	args := []string{"up", "-d"}
 	if conf.QuietPull {
 		args = append(args, "--quiet-pull")
 	}
+
+	composePath, err := c.getCompose(ctx, args...)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:269]:
```suggestion
	commands, err := c.getCompose(ctx, args...)
```

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 11]
```diff
@@ -280,10 +287,17 @@ func (c *Cluster) Down(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	args := []string{"compose", "down"}
+
+	args := []string{"down"}
+	composePath, err := c.getCompose(ctx, args...)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:292]:
```suggestion
	commands, err := c.getCompose(ctx, args...)
```

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 12]
```diff
@@ -384,3 +398,23 @@ func (c *Cluster) ListImages(ctx context.Context, actual bool) ([]string, error)
 		conf.PrometheusImage,
 	}, nil
 }
+
+func (c *Cluster) getCompose(ctx context.Context, args ...string) ([]string, error) {
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:402]:
```suggestion
func (c *Cluster) buildComamnds(ctx context.Context, args ...string) ([]string, error) {
```

**Comment by wzshiming** [comment]
/retitle Download Docker-Compose-v2 automatically if necessary

File: pkg/kwokctl/vars/vars.go [hunk 13]
```diff
@@ -127,6 +129,8 @@ var (
 	// MustKubectlBinary is the binary of kubectl.
 	MustKubectlBinary = "https://dl.k8s.io/release/" + KubeVersion + "/bin/" + runtime.GOOS + "/" + runtime.GOARCH + "/kubectl" + BinSuffix
 
+	MustDockerComposeBinary = "https://github.com/docker/compose/releases/download/" + DockerComposeVersion + "/docker-compose" + "-" + runtime.GOOS + "-" + archAlias(runtime.GOARCH) + BinSuffix
+
```
Comment (wzshiming) [inline pkg/kwokctl/vars/vars.go:133]:
```suggestion
	# DockerComposeBinaryPrefix is the binary of docker-compose.
	DockerComposeBinaryPrefix = "https://github.com/docker/compose/releases/download/" + DockerComposeVersion 
	
	MustDockerComposeBinary = DockerComposeBinaryPrefix + "/docker-compose-" + runtime.GOOS + "-" + archAlias(runtime.GOARCH) + BinSuffix

```

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 14]
```diff
@@ -261,13 +261,21 @@ func (c *Cluster) Up(ctx context.Context) error {
...
-	args := []string{"compose", "up", "-d"}
+
+	args := []string{"up", "-d"}
 	if conf.QuietPull {
 		args = append(args, "--quiet-pull")
 	}
+
+	commands, err := c.buildComamnds(ctx, args...)
+	if err != nil {
+		c.Logger().Printf("Download Compose failed: %v", err)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:272]:
```suggestion
		c.Logger().Printf("Failed to build commands: %v", err)
```

File: pkg/kwokctl/runtime/compose/cluster.go [hunk 15]
```diff
@@ -280,10 +288,17 @@ func (c *Cluster) Down(ctx context.Context) error {
 	if err != nil {
 		return err
 	}
-	args := []string{"compose", "down"}
+
+	args := []string{"down"}
+	commands, err := c.buildComamnds(ctx, args...)
+	if err != nil {
+		c.Logger().Printf("Download Compose failed: %v", err)
```
Comment (wzshiming) [inline pkg/kwokctl/runtime/compose/cluster.go:295]:
```suggestion
		c.Logger().Printf("Failed to build commands: %v", err)
```

**Review by wzshiming — CHANGES_REQUESTED** [review]

File: pkg/kwokctl/vars/vars.go [hunk 16]
```diff
@@ -127,6 +130,11 @@ var (
 	// MustKubectlBinary is the binary of kubectl.
 	MustKubectlBinary = "https://dl.k8s.io/release/" + KubeVersion + "/bin/" + runtime.GOOS + "/" + runtime.GOARCH + "/kubectl" + BinSuffix
 
+	// DockerComposeBinaryPrefix is the binary of docker-compose.
+	DockerComposeBinaryPrefix = getEnv("KWOK_DOCKER_COMPOSE_BINARY_PERFIX", "https://github.com/docker/compose/releases/download/"+DockerComposeVersion)
+
+	MustDockerComposeBinary = DockerComposeBinaryPrefix + "/docker-compose" + "-" + runtime.GOOS + "-" + archAlias(runtime.GOARCH) + BinSuffix
```
Comment (wzshiming) [inline pkg/kwokctl/vars/vars.go:136]: Follow other variables, allow args in `kwokctl create cluster` to modify it

**Comment by wzshiming** [comment]
> Delete the download docker-compose logic from CI for test the patch is working

**Review by wzshiming — CHANGES_REQUESTED** [review]

**Comment by wzshiming** [comment]
> Delete the download docker-compose logic from CI for test the patch is working
You missed that comment

**Comment by wzshiming** [comment]
Fixed by #104
/close

**Pull request closed without merging**